akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala

```diff
@@ -10,18 +10,21 @@ import akka.stream.{ FlowShape, Materializer }
 import akka.stream.stage.GraphStage
 import akka.util.ByteString
 import headers.HttpEncoding
-import akka.stream.scaladsl.{ Sink, Source, Flow }
+import akka.stream.scaladsl.{ Flow, Sink, Source }
 
 import scala.concurrent.Future
 
 trait Decoder {
   def encoding: HttpEncoding
 
-  def decode[T <: HttpMessage](message: T)(implicit mapper: DataMapper[T]): T#Self =
+  def decodeMessage(message: HttpMessage): HttpMessage#Self =
```

Comment:
Kudos to @rkuhn!! I will try this later and try to understand the trickery behind it :)